File: solutions/javascript/zebra-puzzle/zebra-puzzle.js
// @ts-check

/** @typedef {("Norwegian"|"Englishman"|"Spaniard"|"Ukrainian"|"Japanese")} Nationality */

/** @typedef {("red"|"green"|"yellow"|"blue"|"ivory")} Color */

/** @typedef {("dog"|"snail"|"fox"|"horse"|"zebra")} Pet */

/** @typedef {("coffee"|"tea"|"milk"|"orange-juice"|"water")} Beverage */

/** @typedef {("old-gold"|"kools"|"chesterfield"|"lucky-strike"|"parliaments")} CigaretteBrand */

/** @type {Nationality[]} */
const NATIONALITIES = [
  "Norwegian",
  "Englishman",
  "Spaniard",
  "Ukrainian",
  "Japanese",
];

/** @type {Color[]} */
const COLORS = ["red", "green", "yellow", "blue", "ivory"];

/** @type {Pet[]} */
const PETS = ["dog", "snail", "fox", "horse", "zebra"];

/** @type {Beverage[]} */
const BEVERAGES = ["coffee", "tea", "milk", "orange-juice", "water"];

/** @type {CigaretteBrand[]} */
const CIGARETTE_BRANDS = [
  "old-gold",
  "kools",
  "chesterfield",
  "lucky-strike",
  "parliaments",
];

class House {
  /** @type {Nationality} */
  resident;
  /** @type {Color} */
  color;
  /** @type {Pet} */
  pet;
  /** @type {Beverage} */
  beverage;
  /** @type {CigaretteBrand} */
  cigBrand;

  /**
   *
   * @param {Nationality} resident
   * @param {Color} color
   * @param {Pet} pet
   * @param {Beverage} beverage
   * @param {CigaretteBrand} cigBrand
   */
  constructor(resident, color, pet, beverage, cigBrand) {
    this.resident = resident;
    this.color = color;
    this.pet = pet;
    this.beverage = beverage;
    this.cigBrand = cigBrand;
  }

  /**
   * 
   * @param {House} h 
   * @returns {Boolean}
   */
  similarTo(h) {
    return h.resident === this.resident
        || h.color === this.color
        || h.pet === this.pet
        || h.beverage === this.beverage
        || h.cigBrand === this.cigBrand;
  }
}

function* Houses() {
  for (const nationality of NATIONALITIES)
    for (const color of COLORS)
      for (const pet of PETS)
        for (const beverage of BEVERAGES)
          for (const cigBrand of CIGARETTE_BRANDS)
            yield new House(nationality, color, pet, beverage, cigBrand);
}

function* PossibleHouses() {
  for (const house of Houses()) {
    if ((house.resident === "Englishman") !== (house.color === "red")) continue;
    if ((house.resident === "Spaniard") !== (house.pet === "dog")) continue;
    if ((house.beverage === "coffee") !== (house.color === "green")) continue;
    if ((house.resident === "Ukrainian") !== (house.beverage === "tea")) continue;
    if ((house.cigBrand === "old-gold") !== (house.pet === "snail")) continue;
    if ((house.cigBrand === "kools") !== (house.color === "yellow")) continue;
    if ((house.cigBrand === "lucky-strike") !== (house.beverage === "orange-juice")) continue;
    if ((house.cigBrand === "parliaments") !== (house.resident === "Japanese")) continue;
    yield house;
  }
}

/**
 * @param {House[]} houses
 */
function* PossibleHousesExcept(...houses) {
  for (const house of PossibleHouses()) {
    if (houses.some(h => h.similarTo(house)))
      continue;
    yield house;
  }
}

function* Village() {
  for (const h1 of PossibleHouses())
    for (const h2 of PossibleHousesExcept(h1))
      for (const h3 of PossibleHousesExcept(h1, h2))
        for (const h4 of PossibleHousesExcept(h1, h2, h3))
          for (const h5 of PossibleHousesExcept(h1, h2, h3, h4))
        yield [h1, h2, h3, h4, h5];
}

/**
 * @callback Predicate
 * @param {House} a
 * @param {House} b
 * @returns {boolean}
 */

/**
 * @param {Predicate} f
 */
function ambivalent(f) {
  return (/** @type {House} */ a, /** @type {House} */ b) => f(a, b) || f(b, a);
}

/**
 * @type {Predicate}
 */
const chesterFox = ambivalent((a, b) => a.cigBrand === "chesterfield" && b.pet === "fox");

/**
 * @type {Predicate}
 */
const koolsHorse = ambivalent((a, b) => a.cigBrand === "kools" && b.pet === "horse");

/**
 * @type {Predicate}
 */
const norwayBlue = ambivalent((a, b) => a.color === "blue" && b.resident === "Norwegian");

/**
 * @type {Predicate}
 */
function greenIvory(lft, rgt) {
  return lft.color === "ivory" && rgt.color === "green";
}

/**
 * @param {Predicate} f
 * @param {House} h1
 * @param {House} h2
 * @param {House} h3
 * @param {House} h4
 * @param {House} h5
 */
function villageCheck(f, h1, h2, h3, h4, h5) {
  return f(h1, h2) || f(h2, h3) || f(h3, h4) || f(h4, h5);
}

function* Solution() {
  for (const [h1, h2, h3, h4, h5] of Village()) {
    if (h1.resident !== "Norwegian") continue;
    if (h2.color !== "blue") continue;
    if (h3.beverage !== "milk") continue;
    if (h5.color === "ivory") continue;
    if (!villageCheck(greenIvory, h1, h2, h3, h4, h5)) continue;
    if (!villageCheck(chesterFox, h1, h2, h3, h4, h5)) continue;
    if (!villageCheck(koolsHorse, h1, h2, h3, h4, h5)) continue;
    if (!villageCheck(norwayBlue, h1, h2, h3, h4, h5)) continue;
    yield [h1, h2, h3, h4, h5];
  }
}


export class ZebraPuzzle {
  /** @type {House[]} */
  solution;

  constructor() {
    const solution = Solution().next().value;
    if (solution == null) throw new Error("No solution found :(");
    this.solution = solution;
  }

  waterDrinker() {
    return this.solution.find((h) => h.beverage === "water")?.resident;
  }

  zebraOwner() {
    return this.solution.find((h) => h.pet === "zebra")?.resident;
  }

}

File: solutions/scala/zebra-puzzle/src/main/scala/ZebraPuzzle.scala
object ZebraPuzzle {

  sealed trait Resident
  case object Englishman extends Resident
  case object Spaniard extends Resident
  case object Ukrainian extends Resident
  case object Norwegian extends Resident
  case object Japanese extends Resident

  sealed trait Color
  case object Red extends Color
  case object Green extends Color
  case object Yellow extends Color
  case object Blue extends Color
  case object Ivory extends Color

  sealed trait Pet
  case object Dog extends Pet
  case object Snail extends Pet
  case object Fox extends Pet
  case object Horse extends Pet
  case object Zebra extends Pet

  sealed trait Beverage
  case object Coffee extends Beverage
  case object Tea extends Beverage
  case object Milk extends Beverage
  case object OrangeJuice extends Beverage
  case object Water extends Beverage

  sealed trait CigaretteBrand
  case object OldGold extends CigaretteBrand
  case object Kools extends CigaretteBrand
  case object Chesterfield extends CigaretteBrand
  case object LuckyStrike extends CigaretteBrand
  case object Parliaments extends CigaretteBrand
  
  case class House(resident: Resident, color: Color, pet: Pet, beverage: Beverage, cigBrand: CigaretteBrand) {
    def hasEqual(other: House) =
      resident == other.resident ||
      color == other.color ||
      pet == other.pet ||
      beverage == other.beverage ||
      cigBrand == other.cigBrand
  }
  
  case class Solution(waterDrinker: Resident, zebraOwner: Resident)

  val RESIDENTS = Seq(Englishman, Spaniard, Ukrainian, Norwegian, Japanese)
  val COLORS = Seq(Red, Green, Yellow, Blue, Ivory)
  val PETS = Seq(Dog, Snail, Fox, Horse, Zebra)
  val BEVERAGES = Seq(Coffee, Tea, Milk, OrangeJuice, Water)
  val CIGARETTE_BRANDS = Seq(OldGold, Kools, Chesterfield, LuckyStrike, Parliaments)
  
  
  lazy val solve: Solution = {
    val potential = for {
      resident <- RESIDENTS
      color <- COLORS
      pet <- PETS
      beverage <- BEVERAGES
      cigBrand <- CIGARETTE_BRANDS
      if (resident == Englishman) == (color == Red)
      if (resident == Spaniard) == (pet == Dog)
      if (beverage == Coffee) == (color == Green)
      if (resident == Ukrainian) == (beverage == Tea)
      if (cigBrand == OldGold) == (pet == Snail)
      if (cigBrand == Kools) == (color == Yellow)
      if (cigBrand == LuckyStrike) == (beverage == OrangeJuice)
      if (resident == Japanese) == (cigBrand == Parliaments)
    } yield House(resident, color, pet, beverage, cigBrand)
    
    def as(taken: Seq[House]) = potential.view.filterNot(p => taken.exists(_.hasEqual(p)))
    
    def greenIvory(lft: House, rgt: House) = lft.color == Ivory && rgt.color == Green // #6

    def checkChesterFox(a: House, b: House) = a.cigBrand == Chesterfield && b.pet == Fox
    def chesterFox(a: House, b: House) = checkChesterFox(a, b) || checkChesterFox(b, a) // #11
    
    def checkKoolsHorse(a: House, b: House) = a.cigBrand == Kools && b.pet == Horse
    def koolsHorse(a: House, b: House) = checkKoolsHorse(a, b) || checkKoolsHorse(b, a) // #12
    
    def checkNorwayBlue(a: House, b: House) = a.color == Blue && b.resident == Norwegian
    def norwayBlue(a: House, b: House) = checkNorwayBlue(a, b) || checkNorwayBlue(b, a) // #15

    val houses = for {
      h1 <- as(Seq.empty)
      if h1.resident == Norwegian
      h2 <- as(Seq(h1))
      if h2.color == Blue // #15
      h3 <- as(Seq(h1, h2))
      if h3.beverage == Milk // #9
      h4 <- as(Seq(h1, h2, h3))
      h5 <- as(Seq(h1, h2, h3, h4))
      if h5.color != Ivory // #6 (impl)
      if greenIvory(h1, h2) || greenIvory(h2, h3) || greenIvory(h3, h4) || greenIvory(h4, h5) // #6
      if chesterFox(h1, h2) || chesterFox(h2, h3) || chesterFox(h3, h4) || chesterFox(h4, h5) // #11
      if koolsHorse(h1, h2) || koolsHorse(h2, h3) || koolsHorse(h3, h4) || koolsHorse(h4, h5) // #12
      if norwayBlue(h1, h2) || norwayBlue(h2, h3) || norwayBlue(h3, h4) || norwayBlue(h4, h5) // #15
    } yield Seq(h1, h2, h3, h4, h5)
    val found = houses.head
    Solution(found.filter(_.beverage == Water).head.resident, found.filter(_.pet == Zebra).head.resident)
  }
}

